man/smpquery.8
.TH SMPQUERY 8 "March 9, 2011" "OpenIB" "OpenIB Diagnostics"

.SH NAME
smpquery \- query InfiniBand subnet management attributes

.SH SYNOPSIS
.B smpquery
[\-d(ebug)] [\-e(rr_show)] [\-v(erbose)] [\-D(irect)] [\-G(uid)]
[\-x|\-\-extended] [\-C ca_name] [\-P ca_port] [\-t(imeout) timeout_ms]
[\-\-node-name-map node-name-map] [\-V(ersion)] [\-h(elp)]
<op> <dest dr_path|lid|guid> [op params]

.SH DESCRIPTION
.PP
smpquery allows a basic subset of standard SMP queries including the following:
node info, node description, switch info, port info. Fields are displayed in
human readable format.

.SH OPTIONS

.PP
.TP
Current supported operations and their parameters:
        nodeinfo <addr>
        nodedesc <addr>
        portinfo <addr> [<portnum>]     # default port is zero
        switchinfo <addr>
        pkeys <addr> [<portnum>]
        sl2vl <addr> [<portnum>]
        vlarb <addr> [<portnum>]
        guids <addr>
        mlnxextportinfo <addr> [<portnum>]  # default port is zero

.TP
\fB\-x\fR, \fB\-\-extended\fR
Set SMSupportsExtendedSpeeds bit 31 in AttributeModifier
(only impacts PortInfo queries).

.TP
\fB\-\-node\-name\-map\fR <node-name-map>
Specify a node name map.  The node name map file maps GUIDs to more
user friendly names.  See
.B ibnetdiscover(8)
for node name map file format.

.SH COMMON OPTIONS

Most OpenIB diagnostics take the following common flags. The exact list of
supported flags per utility can be found in the usage message and can be shown
using the util_name -h syntax.

# Debugging flags
.PP
\-d      raise the IB debugging level.
        May be used several times (-ddd or -d -d -d).
.PP
\-e      show send and receive errors (timeouts and others)
.PP
\-h      show the usage message
.PP
\-v      increase the application verbosity level.
        May be used several times (-vv or -v -v -v)
.PP
\-V      show the version info.

# Addressing flags
.PP
\-D      use directed path address arguments. The path
        is a comma separated list of out ports.
        Examples:
        "0"             # self port
        "0,1,2,1,4"     # out via port 1, then 2, ...
.PP
\-c      use combined route address arguments. The
        address is a combination of a LID and a direct route path.
        The LID specified is the DLID and the local LID is used
        as the DrSLID.
.PP
\-G      use GUID address argument. In most cases, it is the Port GUID.
        Example:
        "0x08f1040023"
.PP
\-s <smlid>      use 'smlid' as the target lid for SM/SA queries.

# Other common flags:
.PP
\-C <ca_name>    use the specified ca_name.
.PP
\-P <ca_port>    use the specified ca_port.
.PP
\-t <timeout_ms> override the default timeout for the solicited mads.

Multiple CA/Multiple Port Support

When no IB device or port is specified, the port to use is selected
by the following criteria:
.PP
1. the first port that is ACTIVE.
.PP
2. if not found, the first port that is UP (physical link up).

If a port and/or CA name is specified, the user request is
attempted to be fulfilled, and will fail if it is not possible.

.SH EXAMPLES

.PP
smpquery portinfo 3 1        # portinfo by lid, with port modifier
.PP
smpquery -G switchinfo 0x2C9000100D051 1        # switchinfo by guid
.PP
smpquery -D nodeinfo 0       # nodeinfo by direct route
.PP
smpquery -c nodeinfo 6 0,12  # nodeinfo by combined route

.SH SEE ALSO
.BR smpdump (8)

.SH AUTHOR
.TP
Hal Rosenstock
.RI < hal@mellanox.com >


ChangeLog
commit b371f8fa85650b95647e3fec1433067992b8c66a
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu Mar 10 09:44:34 2011 -0600

    Spin release 1.1.1.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit ace467085deb7327522faeb19e20e13d691f66a2
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu Mar 10 09:34:11 2011 -0600

    Flush MA FIFO for on-chip queues.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 7cb87f46c1784c791d29646104ab355ed920747d
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu Mar 10 09:27:57 2011 -0600

    increase the frequency of CIDX-INC updates.
    
    Heavy IMB loads reveal that my current algorithm of doing CIDX-INC
    GTS writes every 1/2 the CQ depth was not sufficient.  This patch
    increases the frequency to 1/16.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 7652d71daadda0c3aba9b96311ea623a1e324d77
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed Sep 15 16:57:04 2010 -0500

    Spin release 1.1.0.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit ed05d22592bc96987635b2c255a99663f286b812
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed Sep 15 16:56:51 2010 -0500

    Support all T4 device IDs.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit b051a2ea9dc7b6ffa08f955e60b331e42fec8672
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed Sep 15 15:54:35 2010 -0500

    Support ABI 0 and 1 versions of the iwarp driver.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 62f1363bb716e2283f9e2b885e5f02a6d5c692d1
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed Sep 15 14:06:17 2010 -0500

    Bump the slots per SQ to 5 to allow up to 128MB fast registers.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit d51e041839254d656cc260dba3f26a7c9d73aa56
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed Sep 15 14:06:16 2010 -0500

    Raw QP implementation
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 8f55539d0b8aea70084edf31687e590cf5a57221
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed Sep 15 14:03:55 2010 -0500

    Add compile-time user mode rdma stats.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 2e3531bbd91d7a6f057b3503132aef5e64adb548
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed Sep 15 13:57:03 2010 -0500

    Support on-chip SQs.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 65bb173b86c2e0a9ff5f877e4d3951cb8daa6e97
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed Sep 15 13:54:44 2010 -0500

    Work around TERM CQE hw bug.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 6bb37d1f72652d4ed0c1972bc673f4418cc7a856
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed Sep 15 13:54:44 2010 -0500

    Zero out padding in iSGLs.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit aad128404c494ec4779055b5e550c2d7de7d5ca9
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu Jul 22 14:58:51 2010 -0500

    Spin release 1.0.4
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 07ac253a8a479fd59fe654baf5a94ffc1de13765
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu Jul 22 14:57:27 2010 -0500

    Do GTS/CIDX_INC writes in the poll path to avoid drops.
    
    With T4 asic, we still see the CIDX_INC overflow/drops.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 3394198639f811c86ab519c43197a9b724435cb0
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu Jul 22 14:56:53 2010 -0500

    Support T420 and T440 cards.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 311dd1c8ec3e38287895053dce52dfef5ed5f0b1
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Fri Jun 11 10:40:53 2010 -0500

    Spin release 1.0.3.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit a62d7cf38f7b6c731486ff969ee7e3a4238ebc3f
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Fri Jun 11 10:37:17 2010 -0500

    Support variable sized work requests.
    
    T4 EQ entries are in multiples of 64B.  Currently the RDMA SQ and RQ
    use fixed sized entries composed of 4 EQ entries for the SQ and 2 EQ
    entries for the RQ.  For optimial latency with small IO, we need to
    change this so the HW only needs to DMA the EQ entries actually used by
    a given work request.
    
    Implementation:
    
    - add wq_pidx counter to track where we are in the EQ.  cidx/pidx are
    used for the sw sq/rq tracking and flow control.
    
    - the variable part of work requests is the SGL.  Add new functions to
    build the SGL and/or immediate data directly in the EQ memory wrapping
    when needed.
    
    - adjust the min burst size for the EQ contexts to 64B.

commit 3be829d117a8ca7957370b09b477fe551de0cea5
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu May 20 15:52:59 2010 -0500

    Spin release 1.0.2.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit dc1ea232e8ece4ba53d2496befad14bd433e2e3e
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu May 20 15:48:37 2010 -0500

    only insert sq qid in lookup tables.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 00c60fd47b71dac42d864662567701e2618ff45d
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu May 20 15:47:37 2010 -0500

    Update some driver limits.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit f25661e19d728a0a33a221c0646a9fce9b160f8b
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu May 20 15:46:38 2010 -0500

    Fix cq arm overflow.
    
    - wrap cq->cqidx_inc based on cq size
    - optimize t4_arm_cq() logic
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>


src/smpquery.c
/*
 * Copyright (c) 2004-2009 Voltaire Inc.  All rights reserved.
 * Copyright (c) 2011 Mellanox Technologies LTD.  All rights reserved.
 *
 * This software is available to you under a choice of one of two
 * licenses.  You may choose to be licensed under the terms of the GNU
 * General Public License (GPL) Version 2, available from the file
 * COPYING in the main directory of this source tree, or the
 * OpenIB.org BSD license below:
 *
 *     Redistribution and use in source and binary forms, with or
 *     without modification, are permitted provided that the following
 *     conditions are met:
 *
 *      - Redistributions of source code must retain the above
 *        copyright notice, this list of conditions and the following
 *        disclaimer.
 *
 *      - Redistributions in binary form must reproduce the above
 *        copyright notice, this list of conditions and the following
 *        disclaimer in the documentation and/or other materials
 *        provided with the distribution.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
 * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
 * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */

#if HAVE_CONFIG_H
#  include <config.h>
#endif				/* HAVE_CONFIG_H */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <getopt.h>
#include <netinet/in.h>

#define __STDC_FORMAT_MACROS
#include <inttypes.h>

#include <infiniband/umad.h>
#include <infiniband/mad.h>
#include <complib/cl_nodenamemap.h>

#include "ibdiag_common.h"

struct ibmad_port *srcport;

typedef char *(op_fn_t) (ib_portid_t * dest, char **argv, int argc);

typedef struct match_rec {
	const char *name, *alias;
	op_fn_t *fn;
	unsigned opt_portnum;
} match_rec_t;

static op_fn_t node_desc, node_info, port_info, switch_info, pkey_table,
    sl2vl_table, vlarb_table, guid_info, mlnx_ext_port_info;

static const match_rec_t match_tbl[] = {
	{"NodeInfo", "NI", node_info},
	{"NodeDesc", "ND", node_desc},
	{"PortInfo", "PI", port_info, 1},
	{"SwitchInfo", "SI", switch_info},
	{"PKeyTable", "PKeys", pkey_table, 1},
	{"SL2VLTable", "SL2VL", sl2vl_table, 1},
	{"VLArbitration", "VLArb", vlarb_table, 1},
	{"GUIDInfo", "GI", guid_info},
	{"MlnxExtPortInfo", "MEPI", mlnx_ext_port_info, 1},
	{0}
};

static char *node_name_map_file = NULL;
static nn_map_t *node_name_map = NULL;
static int extended_speeds = 0;

/*******************************************/
static char *node_desc(ib_portid_t * dest, char **argv, int argc)
{
	int node_type, l;
	uint64_t node_guid;
	char nd[IB_SMP_DATA_SIZE] = { 0 };
	uint8_t data[IB_SMP_DATA_SIZE] = { 0 };
	char dots[128];
	char *nodename = NULL;

	if (!smp_query_via(data, dest, IB_ATTR_NODE_INFO, 0, 0, srcport))
		return "node info query failed";

	mad_decode_field(data, IB_NODE_TYPE_F, &node_type);
	mad_decode_field(data, IB_NODE_GUID_F, &node_guid);

	if (!smp_query_via(nd, dest, IB_ATTR_NODE_DESC, 0, 0, srcport))
		return "node desc query failed";

	nodename = remap_node_name(node_name_map, node_guid, nd);

	l = strlen(nodename);
	if (l < 32) {
		memset(dots, '.', 32 - l);
		dots[32 - l] = '\0';
	} else {
		dots[0] = '.';
		dots[1] = '\0';
	}

	printf("Node Description:%s%s\n", dots, nodename);
	free(nodename);
	return 0;
}

static char *node_info(ib_portid_t * dest, char **argv, int argc)
{
	char buf[2048];
	char data[IB_SMP_DATA_SIZE] = { 0 };

	if (!smp_query_via(data, dest, IB_ATTR_NODE_INFO, 0, 0, srcport))
		return "node info query failed";

	mad_dump_nodeinfo(buf, sizeof buf, data, sizeof data);

	printf("# Node info: %s\n%s", portid2str(dest), buf);
	return 0;
}

static char *port_info(ib_portid_t * dest, char **argv, int argc)
{
	char buf[2300];
	char data[IB_SMP_DATA_SIZE] = { 0 };
	int portnum = 0, orig_portnum;

	if (argc > 0)
		portnum = strtol(argv[0], 0, 0);
	orig_portnum = portnum;
	if (extended_speeds)
		portnum |= 1 << 31;

	if (!smp_query_via(data, dest, IB_ATTR_PORT_INFO, portnum, 0, srcport))
		return "port info query failed";

	mad_dump_portinfo(buf, sizeof buf, data, sizeof data);

	printf("# Port info: %s port %d\n%s", portid2str(dest), orig_portnum, buf);
	return 0;
}

static char *mlnx_ext_port_info(ib_portid_t * dest, char **argv, int argc)
{
	char buf[2300];
	char data[IB_SMP_DATA_SIZE];
	int portnum = 0;

	if (argc > 0)
		portnum = strtol(argv[0], 0, 0);

	if (!smp_query_via(data, dest, IB_ATTR_MLNX_EXT_PORT_INFO, portnum, 0, srcport))
		return "Mellanox ext port info query failed";

	mad_dump_mlnx_ext_port_info(buf, sizeof buf, data, sizeof data);

	printf("# MLNX ext Port info: %s port %d\n%s", portid2str(dest), portnum, buf);
	return 0;
}

static char *switch_info(ib_portid_t * dest, char **argv, int argc)
{
	char buf[2048];
	char data[IB_SMP_DATA_SIZE] = { 0 };

	if (!smp_query_via(data, dest, IB_ATTR_SWITCH_INFO, 0, 0, srcport))
		return "switch info query failed";

	mad_dump_switchinfo(buf, sizeof buf, data, sizeof data);

	printf("# Switch info: %s\n%s", portid2str(dest), buf);
	return 0;
}

static char *pkey_table(ib_portid_t * dest, char **argv, int argc)
{
	uint8_t data[IB_SMP_DATA_SIZE] = { 0 };
	int i, j, k;
	uint16_t *p;
	unsigned mod;
	int n, t, phy_ports;
	int portnum = 0;

	if (argc > 0)
		portnum = strtol(argv[0], 0, 0);

	/* Get the partition capacity */
	if (!smp_query_via(data, dest, IB_ATTR_NODE_INFO, 0, 0, srcport))
		return "node info query failed";

	mad_decode_field(data, IB_NODE_TYPE_F, &t);
	mad_decode_field(data, IB_NODE_NPORTS_F, &phy_ports);
	if (portnum > phy_ports)
		return "invalid port number";

	if ((t == IB_NODE_SWITCH) && (portnum != 0)) {
		if (!smp_query_via(data, dest, IB_ATTR_SWITCH_INFO, 0, 0,
				   srcport))
			return "switch info failed";
		mad_decode_field(data, IB_SW_PARTITION_ENFORCE_CAP_F, &n);
	} else
		mad_decode_field(data, IB_NODE_PARTITION_CAP_F, &n);

	for (i = 0; i < (n + 31) / 32; i++) {
		mod = i | (portnum << 16);
		if (!smp_query_via(data, dest, IB_ATTR_PKEY_TBL, mod, 0,
				   srcport))
			return "pkey table query failed";
		if (i + 1 == (n + 31) / 32)
			k = ((n + 7 - i * 32) / 8) * 8;
		else
			k = 32;
		p = (uint16_t *) data;
		for (j = 0; j < k; j += 8, p += 8) {
			printf
			    ("%4u: 0x%04x 0x%04x 0x%04x 0x%04x 0x%04x 0x%04x 0x%04x 0x%04x\n",
			     (i * 32) + j, ntohs(p[0]), ntohs(p[1]),
			     ntohs(p[2]), ntohs(p[3]), ntohs(p[4]), ntohs(p[5]),
			     ntohs(p[6]), ntohs(p[7]));
		}
	}
	printf("%d pkeys capacity for this port\n", n);

	return 0;
}

static char *sl2vl_dump_table_entry(ib_portid_t * dest, int in, int out)
{
	char buf[2048];
	char data[IB_SMP_DATA_SIZE] = { 0 };
	int portnum = (in << 8) | out;

	if (!smp_query_via(data, dest, IB_ATTR_SLVL_TABLE, portnum, 0, srcport))
		return "slvl query failed";

	mad_dump_sltovl(buf, sizeof buf, data, sizeof data);
	printf("ports: in %2d, out %2d: ", in, out);
	printf("%s", buf);
	return 0;
}

static char *sl2vl_table(ib_portid_t * dest, char **argv, int argc)
{
	uint8_t data[IB_SMP_DATA_SIZE] = { 0 };
	int type, num_ports, portnum = 0;
	int i;
	char *ret;

	if (argc > 0)
		portnum = strtol(argv[0], 0, 0);

	if (!smp_query_via(data, dest, IB_ATTR_NODE_INFO, 0, 0, srcport))
		return "node info query failed";

	mad_decode_field(data, IB_NODE_TYPE_F, &type);
	mad_decode_field(data, IB_NODE_NPORTS_F, &num_ports);
	if (portnum > num_ports)
		return "invalid port number";

	printf("# SL2VL table: %s\n", portid2str(dest));
	printf("#                 SL: |");
	for (i = 0; i < 16; i++)
		printf("%2d|", i);
	printf("\n");

	if (type != IB_NODE_SWITCH)
		return sl2vl_dump_table_entry(dest, 0, 0);

	for (i = 0; i <= num_ports; i++) {
		ret = sl2vl_dump_table_entry(dest, i, portnum);
		if (ret)
			return ret;
	}
	return 0;
}

static char *vlarb_dump_table_entry(ib_portid_t * dest, int portnum, int offset,
				    unsigned cap)
{
	char buf[2048];
	char data[IB_SMP_DATA_SIZE] = { 0 };

	if (!smp_query_via(data, dest, IB_ATTR_VL_ARBITRATION,
			   (offset << 16) | portnum, 0, srcport))
		return "vl arb query failed";
	mad_dump_vlarbitration(buf, sizeof(buf), data, cap * 2);
	printf("%s", buf);
	return 0;
}

static char *vlarb_dump_table(ib_portid_t * dest, int portnum,
			      char *name, int offset, int cap)
{
	char *ret;

	printf("# %s priority VL Arbitration Table:", name);
	ret = vlarb_dump_table_entry(dest, portnum, offset,
				     cap < 32 ? cap : 32);
	if (!ret && cap > 32)
		ret = vlarb_dump_table_entry(dest, portnum, offset + 1,
					     cap - 32);
	return ret;
}

static char *vlarb_table(ib_portid_t * dest, char **argv, int argc)
{
	uint8_t data[IB_SMP_DATA_SIZE] = { 0 };
	int portnum = 0;
	int type, enhsp0, lowcap, highcap;
	char *ret = 0;

	if (argc > 0)
		portnum = strtol(argv[0], 0, 0);

	/* port number of 0 could mean SP0 or port MAD arrives on */
	if (portnum == 0) {
		if (!smp_query_via(data, dest, IB_ATTR_NODE_INFO, 0, 0,
				   srcport))
			return "node info query failed";

		mad_decode_field(data, IB_NODE_TYPE_F, &type);
		if (type == IB_NODE_SWITCH) {
			memset(data, 0, sizeof(data));
			if (!smp_query_via(data, dest, IB_ATTR_SWITCH_INFO, 0,
					   0, srcport))
				return "switch info query failed";
			mad_decode_field(data, IB_SW_ENHANCED_PORT0_F, &enhsp0);
			if (!enhsp0) {
				printf
				    ("# No VLArbitration tables (BSP0): %s port %d\n",
				     portid2str(dest), 0);
				return 0;
			}
			memset(data, 0, sizeof(data));
		}
	}

	if (!smp_query_via(data, dest, IB_ATTR_PORT_INFO, portnum, 0, srcport))
		return "port info query failed";

	mad_decode_field(data, IB_PORT_VL_ARBITRATION_LOW_CAP_F, &lowcap);
	mad_decode_field(data, IB_PORT_VL_ARBITRATION_HIGH_CAP_F, &highcap);

	printf("# VLArbitration tables: %s port %d LowCap %d HighCap %d\n",
	       portid2str(dest), portnum, lowcap, highcap);

	if (lowcap > 0)
		ret = vlarb_dump_table(dest, portnum, "Low", 1, lowcap);

	if (!ret && highcap > 0)
		ret = vlarb_dump_table(dest, portnum, "High", 3, highcap);

	return ret;
}

static char *guid_info(ib_portid_t * dest, char **argv, int argc)
{
	uint8_t data[IB_SMP_DATA_SIZE] = { 0 };
	int i, j, k;
	uint64_t *p;
	unsigned mod;
	int n;

	/* Get the guid capacity */
	if (!smp_query_via(data, dest, IB_ATTR_PORT_INFO, 0, 0, srcport))
		return "port info failed";
	mad_decode_field(data, IB_PORT_GUID_CAP_F, &n);

	for (i = 0; i < (n + 7) / 8; i++) {
		mod = i;
		if (!smp_query_via(data, dest, IB_ATTR_GUID_INFO, mod, 0,
				   srcport))
			return "guid info query failed";
		if (i + 1 == (n + 7) / 8)
			k = ((n + 1 - i * 8) / 2) * 2;
		else
			k = 8;
		p = (uint64_t *) data;
		for (j = 0; j < k; j += 2, p += 2) {
			printf("%4u: 0x%016" PRIx64 " 0x%016" PRIx64 "\n",
			       (i * 8) + j, ntohll(p[0]), ntohll(p[1]));
		}
	}
	printf("%d guids capacity for this port\n", n);

	return 0;
}

static op_fn_t *match_op(char *name)
{
	const match_rec_t *r;
	unsigned len = strlen(name);
	for (r = match_tbl; r->name; r++)
		if (!strncasecmp(r->name, name, len) ||
		    (r->alias && !strncasecmp(r->alias, name, len)))
			return r->fn;
	return NULL;
}

static int process_opt(void *context, int ch, char *optarg)
{
	switch (ch) {
	case 1:
		node_name_map_file = strdup(optarg);
		break;
	case 'c':
		ibd_dest_type = IB_DEST_DRSLID;
		break;
	case 'x':
		extended_speeds = 1;
		break;
	default:
		return -1;
	}
	return 0;
}

int main(int argc, char **argv)
{
	char usage_args[1024];
	int mgmt_classes[3] =
	    { IB_SMI_CLASS, IB_SMI_DIRECT_CLASS, IB_SA_CLASS };
	ib_portid_t portid = { 0 };
	char *err;
	op_fn_t *fn;
	const match_rec_t *r;
	int n;

	const struct ibdiag_opt opts[] = {
		{"combined", 'c', 0, NULL,
		 "use Combined route address argument"},
		{"node-name-map", 1, 1, "<file>", "node name map file"},
		{"extended", 'x', 0, NULL, "use extended speeds"},
		{0}
	};
	const char *usage_examples[] = {
		"portinfo 3 1\t\t\t\t# portinfo by lid, with port modifier",
		"-G switchinfo 0x2C9000100D051 1\t# switchinfo by guid",
		"-D nodeinfo 0\t\t\t\t# nodeinfo by direct route",
		"-c nodeinfo 6 0,12\t\t\t# nodeinfo by combined route",
		NULL
	};

	n = sprintf(usage_args, "<op> <dest dr_path|lid|guid> [op params]\n"
		    "\nSupported ops (and aliases, case insensitive):\n");
	for (r = match_tbl; r->name; r++) {
		n += snprintf(usage_args + n, sizeof(usage_args) - n,
			      "  %s (%s) <addr>%s\n", r->name,
			      r->alias ? r->alias : "",
			      r->opt_portnum ? " [<portnum>]" : "");
		if (n >= sizeof(usage_args))
			exit(-1);
	}

	ibdiag_process_opts(argc, argv, NULL, NULL, opts, process_opt,
			    usage_args, usage_examples);

	argc -= optind;
	argv += optind;

	if (argc < 2)
		ibdiag_show_usage();

	if (!(fn = match_op(argv[0])))
		IBERROR("operation '%s' not supported", argv[0]);

	srcport = mad_rpc_open_port(ibd_ca, ibd_ca_port, mgmt_classes, 3);
	if (!srcport)
		IBERROR("Failed to open '%s' port '%d'", ibd_ca, ibd_ca_port);

	node_name_map = open_node_name_map(node_name_map_file);

	if (ibd_dest_type != IB_DEST_DRSLID) {
		if (ib_resolve_portid_str_via(&portid, argv[1], ibd_dest_type,
					      ibd_sm_id, srcport) < 0)
			IBERROR("can't resolve destination port %s", argv[1]);
		if ((err = fn(&portid, argv + 2, argc - 2)))
			IBERROR("operation %s: %s", argv[0], err);
	} else {
		char concat[64];

		memset(concat, 0, 64);
		snprintf(concat, sizeof(concat), "%s %s", argv[1], argv[2]);
		if (ib_resolve_portid_str_via(&portid, concat, ibd_dest_type,
					      ibd_sm_id, srcport) < 0)
			IBERROR("can't resolve destination port %s", concat);
		if ((err = fn(&portid, argv + 3, argc - 3)))
			IBERROR("operation %s: %s", argv[0], err);
	}
	close_node_name_map(node_name_map);
	mad_rpc_close_port(srcport);
	exit(0);
}
